### PR TITLE
Add guide: Detect anomalies in nodes and applications with Netdata

### PR DESCRIPTION
**Pull request by joelhans**
##### Summary

The first part of a 3-part guide that walks users through setting up the anomalies collector using Nginx as an example framework. Part 2 will cover how these metrics/charts/alarms propagate through the Netdata universe, and part 3 will cover advanced custom models.

##### Component Name

area/docs

### CODE REVIEW

**Review by aabatangle — DISMISSED** [review]

File: docs/guides/monitor/anomaly-detection.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+# Detect anomalies in systems and applications
+
+Beginning with v1.27, the [open-source Netdata Agent](https://github.com/netdata/netdata) is capable of unsupervised
+[anomaly detection](https://en.wikipedia.org/wiki/Anomaly_detection) with machine learning (ML). As with all things
+Netdata, the anomalies collector comes with preconfigured alarms and instant visualizations that require no query
+languages or organizing metrics. You configure the collector to look at specific charts, and it handles the rest.
+
+Netdata's implementation uses a handful of functions in the [Python Outlier Detection (PyOD)
+library](https://github.com/yzhao062/pyod/tree/master), which first runs a `train` function that learns what "normal"
+looks like on your node and creates an ML model, then utilizes the `predict_proba()` and `predict()` functions to
```
Comment (andrewm4894) [inline docs/guides/monitor/anomaly-detection.md:17]:
```suggestion
looks like on your node and creates an ML model for each chart, then utilizes the `predict_proba()` and `predict()` functions to
```

File: docs/guides/monitor/anomaly-detection.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+visualize how anomalous certain metrics are.
```
Comment (andrewm4894) [inline docs/guides/monitor/anomaly-detection.md:18]:
```suggestion
visualize how anomalous certain charts are.
```

**Review by andrewm4894 — CHANGES_REQUESTED** [review]

**Review by aabatangle — APPROVED** [review]

File: collectors/python.d.plugin/anomalies/README.md [hunk 3]
```diff
@@ -1,15 +1,16 @@
 <!--
 title: "Anomalies"
+description: "Use ML-driven anomaly detection to narrow your focus to only affected metrics and services/processes on your node to shorten root cause analysis."
 custom_edit_url: https://github.com/netdata/netdata/edit/master/collectors/python.d.plugin/anomalies/README.md
 -->
 
-# Anomalies: ML-driven anomaly detection for the Netdata Agent
```
Comment (jlbriston) [inline collectors/python.d.plugin/anomalies/README.md:6]: I think the more descriptive this can be the better since anomalies seem to be a wide-ranging term

File: docs/guides/monitor/anomaly-detection.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+[anomaly detection](https://en.wikipedia.org/wiki/Anomaly_detection) with machine learning (ML). As with all things
+Netdata, the anomalies collector comes with preconfigured alarms and instant visualizations that require no query
+languages or organizing metrics. You configure the collector to look at specific charts, and it handles the rest.
+
+Netdata's implementation uses a handful of functions in the [Python Outlier Detection (PyOD)
+library](https://github.com/yzhao062/pyod/tree/master), which first runs a `train` function that learns what "normal"
+looks like on your node and creates an ML model for each chart, then utilizes the `predict_proba()` and `predict()` functions to
+visualize how anomalous certain charts are.
+
+All these metrics and alarms are available for centralized monitoring in [Netdata Cloud](https://app.netdata.cloud), if
```
Comment (jlbriston) [inline docs/guides/monitor/anomaly-detection.md:20]:
```suggestion
All these metrics and alarms are available for centralized monitoring in [Netdata Cloud](https://app.netdata.cloud).
```

File: docs/guides/monitor/anomaly-detection.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+you choose to sign up and claim your nodes, which empowers you to run tailored anomaly detection on every node in your
```
Comment (jlbriston) [inline docs/guides/monitor/anomaly-detection.md:21]:
```suggestion
If you choose to sign up for Cloud and claim your nodes, you will have the ability to run tailored anomaly detection on every node in your
```

File: docs/guides/monitor/anomaly-detection.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+In `python.d.conf` file, search for the `anomalies` line. If the line exists, set the value to `yes`. Add the line
+yourself if it doesn't already exist. Either way, the final result should look like:
+
+```conf
+anomalies: yes
+```
+
+[Restart the Agent](/docs/configure/start-stop-restart.md) with `sudo systemctl restart netdata` to start up the
+anomalies collector. By default, the model training process runs every 30 minutes, and uses the previous 4 hours of
+metrics to establish a baseline for health and performance across the default charts to include.
```
Comment (jlbriston) [inline docs/guides/monitor/anomaly-detection.md:77]:
```suggestion
metrics to establish a baseline for health and performance across the default charts included.
```

File: docs/guides/monitor/anomaly-detection.md [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+`web_log_nginx` chart, and specifically the `apps.cpu` and `apps.mem` charts.
+
+![The anomalies collector chart with many
+dimensions](https://user-images.githubusercontent.com/1153921/102813877-db5e4880-4386-11eb-8040-d7a1d7a476bb.png)
+
+### Remove some metrics from anomaly detection
+
+As you can see in the above screenshot, this node is now looking for anomalies in many places. The result is a single
+`anomalies_local.probability` chart with more than twenty dimensions, some of which the dashboard hides at the bottom of
+a scroll-able area. In addition, training and analyzing the anomaly collector on this many charts might require more CPU
```
Comment (jlbriston) [inline docs/guides/monitor/anomaly-detection.md:127]:
```suggestion
a scroll-able area. In addition, training and analyzing the anomaly collector on these many charts might require more CPU
```

File: docs/guides/monitor/anomaly-detection.md [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+volume of requests/responses, not, for example, which type of 4xx response a user might receive.
+
+```conf
+    charts_to_exclude: 'web_log_nginx.excluded_requests,web_log_nginx.responses_by_status_code_class,web_log_nginx.status_code_class_2xx_responses,web_log_nginx.status_code_class_4xx_responses,web_log_nginx.current_poll_uniq_clients,web_log_nginx.requests_by_http_method,web_log_nginx.requests_by_http_version,web_log_nginx.requests_by_ip_proto'
+```
+
+![The anomalies collector with less
+dimensions](https://user-images.githubusercontent.com/1153921/102820642-d69f9180-4392-11eb-91c5-d3d166d40105.png)
+
+Apply the ideas behind the collector's regex and excluding settings to any other
```
Comment (jlbriston) [inline docs/guides/monitor/anomaly-detection.md:146]:
```suggestion
Apply the ideas behind the collector's regex and exclude settings to any other
```

File: docs/guides/monitor/anomaly-detection.md [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+[system](/docs/collect/system-metrics.md), [container](/docs/collect/container-metrics.md), or
+[application](/docs/collect/application-metrics.md) metrics you want to detect anomalies for.
+
+## What's next?
+
+Now that you know how to set up unsupervised anomaly detection in the Netdata Agent, using an Nginx web server as an
+example, time to apply that knowledge to other mission-critical parts of your infrastructure. If you're not sure what to
```
Comment (jlbriston) [inline docs/guides/monitor/anomaly-detection.md:153]:
```suggestion
example, it's time to apply that knowledge to other mission-critical parts of your infrastructure. If you're not sure what to
```

File: docs/guides/monitor/anomaly-detection.md [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+For a more user-friendly anomaly detection experience, try out the [Metric
+Correlations](https://learn.netdata.cloud/docs/cloud/insights/metric-correlations) feature in Netdata Cloud. Metric
+Correlations runs only at your requests, removing unrelated charts from the dashboard to help you focus on root cause
+analysis.
+
+Stay tuned for the next two parts of this guide, which provide more real-world context for the anomalies collector.
+First, maximize the immediate value you get from anomaly detection by tracking preconfigured alarms, visualizing
+anomalies in charts, and building a new dashboard tailored to your applications. Then, learn about creating custom ML
+models, which help you holistically monitor an application or service by monitor anomalies across a _cluster of charts_.
```
Comment (jlbriston) [inline docs/guides/monitor/anomaly-detection.md:165]:
```suggestion
models, which help you holistically monitor an application or service by monitoring anomalies across a _cluster of charts_.
```

**Review by jlbriston — APPROVED** [review]

File: docs/guides/monitor/anomaly-detection.md [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+In `python.d.conf` file, search for the `anomalies` line. If the line exists, set the value to `yes`. Add the line
+yourself if it doesn't already exist. Either way, the final result should look like:
+
+```conf
+anomalies: yes
+```
+
+[Restart the Agent](/docs/configure/start-stop-restart.md) with `sudo systemctl restart netdata` to start up the
+anomalies collector. By default, the model training process runs every 30 minutes, and uses the previous 4 hours of
+metrics to establish a baseline for health and performance across the default included charts.
```
Comment (andrewm4894) [inline docs/guides/monitor/anomaly-detection.md:77]: Could also maybe add a note here like

> It might take 30-60 seconds for the anomaly collector to finish its initial training and have enough data to start generating anomaly scores. You might need to refresh your Netdata browser session too for the new Anomalies section to appear on the charts menu.

Not sure if "charts menu" is correct term. Point being once you enable it and restart netdata it might take up to 60 seconds to appear.

File: docs/guides/monitor/anomaly-detection.md [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+Open `python.d/anomalies.conf` with `edit-conf`.
+
+```bash
+sudo ./edit-config python.d/anomalies.conf
+```
+
+The file contains many user-configurable settings with sane defaults. Here are some important settings that don't
+involve tweaking the behavior of the ML training itself.
+
+- `charts_regex`: Which charts to train against and include in anomaly detection.
```
Comment (andrewm4894) [inline docs/guides/monitor/anomaly-detection.md:90]:
```suggestion
- `charts_regex`: Which charts to train models for and run anomaly detection on.
```

Comment (andrewm4894) [reply docs/guides/monitor/anomaly-detection.md:90]: Just trying to make it clear it's a model for each chart as opposed to one giant model for all the charts as i'd say people could get confused about that. Not sure if my words best way but just trying to be explicit everywhere that its a model per chart. Kinda easier to get your head around then i think.

File: docs/guides/monitor/anomaly-detection.md [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+- `charts_to_exclude`: Specific charts, selected by the regex in `charts_regex`, to exclude.
+- `train_every_n`: How often to train the ML models.
+- `train_n_secs`: The number of historical metrics to train ML models on. The default is 4 hours, but if your node
```
Comment (andrewm4894) [inline docs/guides/monitor/anomaly-detection.md:93]:
```suggestion
- `train_n_secs`: The number of historical observations to train each model on. The default is 4 hours, but if your node
```

File: docs/guides/monitor/anomaly-detection.md [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+  doesn't have historical metrics going back that far, consider [changing the metrics retention
+  policy](/docs/store/change-metrics-storage.md).
+- `custom_models`: A way to define custom models that you want anomaly probabilities for, including multi-node or
+  streaming setups.
+
```
Comment (andrewm4894) [inline docs/guides/monitor/anomaly-detection.md:98]: Could be worth adding a little note in to warn them against doing something silly like trying to set a huge train_n_secs and/or include all charts. That will cause pain. So maybe something like.

> Note that if you set `charts_regex` to cover a lot of charts and/or set `train_n_secs` to be a very large number, then this will impact the resources and time required to train the models. The performance implications here will depend on each node so its best to make incremental changes if you want to include more or different charts and adjust the volume of data used to train the model. `train_max_n` might be useful too here as it will cap via sampling the number of observations actually used to train on.

Comment (andrewm4894) [reply docs/guides/monitor/anomaly-detection.md:98]: A bit wordy, But just sort of something to stop people blowing up their node is kinda where i'm coming from (without wanting to scare them out of using it either).

**Review by andrewm4894 — CHANGES_REQUESTED** [review]
Made some suggestions, all at your discretion, nothing major. 

Love this! like the idea of trying to do a practical example using something like nginx. 

Just as an fyi - there is also some default alarms set in here: 

https://github.com/netdata/netdata/blob/master/health/health.d/anomalies.conf

Mainly just warnings and not at all tuned or scientific or anything yet - really just a stake in the ground to see what people think - too noisy or useful at all. 

We could consider working that into the guide in some way maybe. Mainly as i'm interested to hear from people what works and what doesn't.

**Review by aabatangle — APPROVED** [review]

**Review by andrewm4894 — APPROVED** [review]
Lgtm
